### ISSUE TITLE
ImportError running test cases

### ISSUE THREAD

**Issue by yfche** [GAA-UAM/fda#78]
Hi developer,

Sorry if my question is more like seeking help rather than contributing to the package development. I am trying to run your fda package locally while I could not build it successfully by running into:
**ImportError: dynamic module does not define module export function (PyInit_optimum_reparam_extension)**,  and error: 
**import optimum_reparam_extension**.
I wish this package performs similar function as the fda package in R. I am brand new to Github, so all I did after downloading the source code from Github using "git" command is installing all the requirements in requirements.txt and building the project by running "**python3 setup.py <operations>**". For <operations>, I ran "build", "install", "bdist" and "check". If it ever helps, I am using Python 3.6.4 on a Mac Mojave 10.14. It would be more than great to get your help on where might be going wrong. I indeed have no idea if I should expect this "optimum_reparam_extension" file to be originally in your package, or it should be something generated after I build it myself. Great many thanks!

**Comment by pablomm** [comment]
Hi @yfche, thank you for contributing by reporting this bug.

It seems to be a problem with the compilation of the cython extensions of the package. I've been looking for similar problems and maybe the problem could be because the extension name is not the same as the cython file name, some mac users have had similar problems for this reason in other packages. I also use a Mac Mojave 10.4, but for some reason I haven't received this error. 

I have added a new branch `feature/fix-cython-build` with some changes to fix this problem, changing the name of the extension and other minor changes. I would really appreciate it if you could try if these changes solve it, installing this branch and checking if the problem is fixed. With the following commands you should be able to do it.

```bash
git clone https://github.com/GAA-UAM/fda.git
cd fda
git checkout feature/fix-cython-build
python3 setup.py install
```
After that the package should be imported and work properly. The script ` tests/test_elastic.py` should run correctly with the following output

```bash
python3 tests/test_elastic.py 

............
----------------------------------------------------------------------
Ran 12 tests in 1.209s

OK
```

If after this you still have the same problem, attach the output of the command `setup.py install`, to continue looking for what could be the problem.

**Comment by yfche** [comment]
Hi developer,

Thanks for the instruction! Following your instruction, I can run tests/test_elastic.py correctly with the same output you presented.

I tested all the tests under the tests/ folder. I got two failed test cases for test_basis_evaluation.py and test_magnitude_shape.py, with the output error attached as attached below. Do you know if there is problem with my local python version, or something else wrong? If it helps, my python version is:

yifengs-Mac-Pro:fda yfche$ python3
Python 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 12:04:33)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>

yifengs-Mac-Pro:fda yfche$ which python3
/anaconda3/bin/python3

Error for test_basis_evaluation.py:

yifengs-Mac-Pro:fda yfche$ python3 tests/test_basis_evaluation.py

/anaconda3/lib/python3.6/importlib/_bootstrap_external.py:426: ImportWarning: Not importing directory /anaconda3/lib/python3.6/site-packages/google: missing __init__
  _warnings.warn(msg.format(portions[0]), ImportWarning)
/anaconda3/lib/python3.6/importlib/_bootstrap_external.py:426: ImportWarning: Not importing directory /anaconda3/lib/python3.6/site-packages/mpl_toolkits: missing __init__
  _warnings.warn(msg.format(portions[0]), ImportWarning)
.F.........................
======================================================================
FAIL: test_evaluation_composed_bspline (__main__.TestBasisEvaluationBSpline)
Test the evaluation of FDataBasis the a matrix of times instead of
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_basis_evaluation.py", line 351, in test_evaluation_composed_bspline
    aligned_evaluation=False)[0])
  File "/anaconda3/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 855, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/anaconda3/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not equal

(mismatch 50.0%)
 x: array([0.000782, 0.569612])
 y: array([0.000782, 0.569612])

----------------------------------------------------------------------
Ran 27 tests in 0.594s

FAILED (failures=1)

Error for test_magnitude_shape.py:

yifengs-Mac-Pro:fda yfche$ python3 tests/test_magnitude_shape.py

./anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/anaconda3/lib/python3.6/importlib/_bootstrap_external.py:426: ImportWarning: Not importing directory /anaconda3/lib/python3.6/site-packages/google: missing __init__
  _warnings.warn(msg.format(portions[0]), ImportWarning)
/anaconda3/lib/python3.6/importlib/_bootstrap_external.py:426: ImportWarning: Not importing directory /anaconda3/lib/python3.6/site-packages/mpl_toolkits: missing __init__
  _warnings.warn(msg.format(portions[0]), ImportWarning)
/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py:875: DeprecationWarning: builtin type EagerTensor has no __module__ attribute
  EagerTensor = c_api.TFE_Py_InitEagerTensor(_EagerTensorBase)
/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:45: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
  if d.decorator_argspec is not None), _inspect.getargspec(target))
/anaconda3/lib/python3.6/site-packages/tensorflow/python/keras/_impl/keras/backend.py:4622: ResourceWarning: unclosed file <_io.TextIOWrapper name='/Users/yfche/.keras/keras.json' mode='r' encoding='UTF-8'>
  _config = json.load(open(_config_path))
/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/anaconda3/lib/python3.6/site-packages/tensorflow/python/util/tf_inspect.py:45: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
  if d.decorator_argspec is not None), _inspect.getargspec(target))
F
======================================================================
FAIL: test_magnitude_shape_plot (__main__.TestMagnitudeShapePlot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_magnitude_shape.py", line 94, in test_magnitude_shape_plot
    0, 0, 0, 0, 1]))
  File "/anaconda3/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 963, in assert_array_almost_equal
    precision=decimal)
  File "/anaconda3/lib/python3.6/site-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Arrays are not almost equal to 6 decimals

(mismatch 8.57142857142857%)
 x: array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1])
 y: array([0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1])

----------------------------------------------------------------------
Ran 2 tests in 2.166s

FAILED (failures=1)

Great many thanks for being helpful!

Best regards,
Yifeng




On Apr 15, 2019, at 7:51 PM, Pablo Marcos <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @yfche<https://github.com/yfche>, thank you for contributing by reporting this bug.

It seems to be a problem with the compilation of the cython extensions of the package. I've been looking for similar problems and maybe the problem could be because the extension name is not the same as the cython file name, some mac users have had similar problems for this reason in other packages. I also use a Mac Mojave 10.4, but for some reason I haven't received this error.

I have added a new branch feature/fix-cython-build with some changes to fix this problem, changing the name of the extension and other minor changes. I would really appreciate it if you could try if these changes solve it, installing this branch and checking if the problem is fixed. With the following commands you should be able to do it.

git clone https://github.com/GAA-UAM/fda.git
cd fda
git checkout feature/fix-cython-build
python3 setup.py install

After that the package should be imported and work properly. The script tests/test_elastic.py should run correctly with the following output

python3 tests/test_elastic.py

............
----------------------------------------------------------------------
Ran 12 tests in 1.209s

OK

If after this you still have the same problem, attach the output of the command setup.py install, to continue looking for what could be the problem.

**Comment by pablomm** [comment]
Thank you very much for these contributions @yfche.

I created a pull request (#80) to merge the ImportError solution into develop.

Regarding the tests that are failing, there is a problem in the test `magnitud_shape_plot.py` (issue #77). It has nothing to do with your version of python or numpy. The pull request #79 fix it, it will merge soon into develop.

The problem of `test_basis_evaluation.py` is due to the use of the exact comparison between floating points. You probably have a different version of python/numpy. It's a problem of how the test was done.
 I changed the tests to avoid this bug (Pull request #81).
